Tests/Source/BoxTester.cpp

```diff
@@ -0,0 +1,232 @@
+#include <memory>
+#include <string>
+#include <catch.hpp>
+#include "Rover/Box.hpp"
+#include "Rover/Constant.hpp"
+#include "Rover/Generator.hpp"
+#include "Rover/Lift.hpp"
+#include "Rover/Range.hpp"
+
+using namespace Rover;
+using namespace std::string_literals;
+
+TEST_CASE("test_values", "[Box]") {
+  SECTION("Constant.") {
+    auto c = Constant(5);
+    auto b = Box(c);
+    static_assert(std::is_same_v<std::decay_t<decltype(b)>, Box<int>>);
+    REQUIRE(generate(b) == 5);
+  }
+  SECTION("Range.") {
+    auto r = Range(Constant(5), Constant(10));
+    auto b = Box(r);
+    static_assert(std::is_same_v<std::decay_t<decltype(b)>, Box<int>>);
+    for(int i = 0; i < 100; ++i) {
+      auto result = generate(b);
+      REQUIRE(result >= 5);
+      REQUIRE(result <= 10);
+    }
+  }
+}
+
+TEST_CASE("test_raw_pointers", "[Box]") {
+  SECTION("Constant.") {
+    auto c = new Constant(5);
+    auto b = Box(c);
+    static_assert(std::is_same_v<std::decay_t<decltype(b)>, Box<int>>);
+    REQUIRE(generate(b) == 5);
+  }
+  SECTION("Range.") {
+    auto r = new Range(Constant(5), Constant(10));
+    auto b = Box(r);
+    static_assert(std::is_same_v<std::decay_t<decltype(b)>, Box<int>>);
+    for(int i = 0; i < 100; ++i) {
+      auto result = generate(b);
+      REQUIRE(result >= 5);
+      REQUIRE(result <= 10);
+    }
+  }
+}
+
+TEST_CASE("test_unique_pointers", "[Box]") {
+  SECTION("Constant.") {
+    auto c = std::make_unique<Constant<int>>(5);
+    auto b = Box(std::move(c));
+    static_assert(std::is_same_v<std::decay_t<decltype(b)>, Box<int>>);
+    REQUIRE(generate(b) == 5);
+  }
+  SECTION("Range.") {
+    auto r = std::make_unique<Range<Constant<int>, Constant<int>, void>>(5,
+        10);
+    auto b = Box(std::move(r));
+    static_assert(std::is_same_v<std::decay_t<decltype(b)>, Box<int>>);
+    for(int i = 0; i < 100; ++i) {
+      auto result = generate(b);
+      REQUIRE(result >= 5);
+      REQUIRE(result <= 10);
+    }
+  }
+}
+
+TEST_CASE("test_shared_pointers", "[Box]") {
+  SECTION("Constant.") {
+    auto c = std::make_shared<Constant<int>>(5);
+    auto b = Box(c);
+    static_assert(std::is_same_v<std::decay_t<decltype(b)>, Box<int>>);
+    REQUIRE(generate(b) == 5);
+  }
+  SECTION("Range.") {
+    auto r = std::make_shared<Range<Constant<int>, Constant<int>, void>>(5,
+        10);
+    auto b = Box(r);
+    static_assert(std::is_same_v<std::decay_t<decltype(b)>, Box<int>>);
+    for(int i = 0; i < 100; ++i) {
+      auto result = generate(b);
+      REQUIRE(result >= 5);
+      REQUIRE(result <= 10);
+    }
+  }
+}
+
+TEST_CASE("test_dependent_generators", "[Box]") {
+  SECTION("FlatRanges") {
+    for(int i = 0; i < 100; ++i) {
+      auto r1 = Box(std::make_shared<Range<Constant<int>, Constant<int>,
```

Comment:
These tests fail due to  the Evaluator clashes. Should work once the fix is merged. Will create a separate issue for that and Range copy.

Reply:
That's great, make a case in Fogbugz/Manuscript, assign it to me and then I'll assign it back to you.